kafka-client/src/main/java/dev/responsive/kafka/internal/db/rs3/RS3KVFlushManager.java

```diff
@@ -0,0 +1,217 @@
+import java.util.HashMap;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import org.apache.kafka.common.utils.Bytes;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class RS3KVFlushManager extends KVFlushManager {
+  private static final Logger LOG = LoggerFactory.getLogger(RS3KVFlushManager.class);
+
+  private final UUID storeId;
+  private final RS3Client rs3Client;
+  private final LssId lssId;
+  private final RS3KVTable table;
+  private final HashMap<Integer, Optional<Long>> writtenOffsets;
+  private final int kafkaPartition;
+  private final PssPartitioner pssPartitioner;
+  private final HashMap<Integer, RS3KVWriter> writers = new HashMap<>();
+
+  public RS3KVFlushManager(
+      final UUID storeId,
+      final RS3Client rs3Client,
+      final LssId lssId,
+      final RS3KVTable table,
+      final HashMap<Integer, Optional<Long>> writtenOffsets,
+      final int kafkaPartition,
+      final PssPartitioner pssPartitioner
+  ) {
+    this.storeId = Objects.requireNonNull(storeId);
+    this.rs3Client = Objects.requireNonNull(rs3Client);
+    this.lssId = Objects.requireNonNull(lssId);
+    this.table = Objects.requireNonNull(table);
+    this.writtenOffsets = Objects.requireNonNull(writtenOffsets);
+    this.kafkaPartition = kafkaPartition;
+    this.pssPartitioner = Objects.requireNonNull(pssPartitioner);
+  }
+
+  @Override
+  public String tableName() {
+    return table.name();
+  }
+
+  @Override
+  public TablePartitioner<Bytes, Integer> partitioner() {
+    return new TablePartitioner<Bytes, Integer>() {
+      @Override
+      public Integer tablePartition(int kafkaPartition, Bytes key) {
+        return pssPartitioner.pss(key.get(), kafkaPartition);
+      }
+
+      @Override
+      public Integer metadataTablePartition(int kafkaPartition) {
+        throw new UnsupportedOperationException();
+      }
+    };
+  }
+
+  @Override
+  public RemoteWriter<Bytes, Integer> createWriter(
+      final Integer pssId,
+      final long consumedOffset
+  ) {
+    if (writers.containsKey(pssId)) {
+      throw new IllegalStateException("already created writer for pss " + pssId);
+    }
+    final var maybeWrittenOffset = writtenOffsets.get(pssId);
+    if (maybeWrittenOffset.isPresent()) {
+      final var writtenOffset = maybeWrittenOffset.get();
+      if (writtenOffset >= consumedOffset) {
+        // This can happen after a restart, during restore because we restore from
+        // the earliest written offset over all pss. For most pss we are restoring
+        // from an offset earlier than the pss's last written offset. In that case,
+        // we wait until we have crossed the written offset for the pss before sending
+        // actual records.
+        // TODO: we should be stricter when checking this case
+        return new NoopWriter(kafkaPartition);
```

Comment:
No, the `writers` map is strictly internal to this class and is just used to record the written offsets after flush completes. By excluding it here we just don't include the writer in the set we use to update the written offsets (because we aren't actually writing anything for this pss for this batch)